File: java/Samples/Biometrics/CS/AbisSampleCS/SubjectEditor/CaptureFacePage.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Linq;
using System.Threading;
using System.Windows.Forms;
using Neurotec.Biometrics;
using Neurotec.Biometrics.Client;
using Neurotec.Images;

namespace Neurotec.Samples
{
	public partial class CaptureFacePage : Neurotec.Samples.PageBase
	{
		#region Public constructor

		public CaptureFacePage()
		{
			InitializeComponent();
		}

		#endregion

		#region Private fields

		private NSubject _subject;
		private NBiometricClient _biometricClient;
		private ManualResetEvent _isIdle = new ManualResetEvent(true);
		private NFace _currentBiometric = null;
		private NSubject _newSubject;
		private bool _isExtractStarted = false;
		private int _sessionId = -1;
		private string _titlePrefix = string.Empty;

		#endregion

		#region Public overridden methods

		public override void OnNavigatedTo(params object[] args)
		{
			if (args == null || args.Length != 2) throw new ArgumentException("args");

			_subject = (NSubject)args[0];
			_biometricClient = (NBiometricClient)args[1];
			if (_subject == null || _biometricClient == null) throw new ArgumentException("args");

			_sessionId = _subject.Faces.Count > 0 ? _subject.Faces.Max(x => x.SessionId) + 1 : 0;
			_newSubject = new NSubject();
			_biometricClient.PropertyChanged += OnBiometricClientPropertyChanged;
			_biometricClient.CurrentBiometricCompletedTimeout = 5000;
			_biometricClient.CurrentBiometricCompleted += OnCurrentBiometricCompleted;

			subjectTreeControl.Visible = false;
			icaoWarningView.Visible = false;
			OnFaceCaptureDeviceChanged();
			ToggleRadioButton();
			lblStatus.Visible = false;
			generalizationView.Visible = false;
			faceView.MirrorHorizontally = chbMirrorHorizontally.Checked = SettingsManager.FacesMirrorHorizontally;

			subjectTreeControl.Subject = _newSubject;

			base.OnNavigatedTo(args);
		}

		public override void OnNavigatingFrom()
		{
			if (IsBusy()) LongActionDialog.ShowDialog(this, "Finish current action ...", CancelAndWait);
			_biometricClient.PropertyChanged -= OnBiometricClientPropertyChanged;
			_biometricClient.CurrentBiometricCompleted -= OnCurrentBiometricCompleted;
			_biometricClient.CurrentBiometricCompletedTimeout = 0;
			subjectTreeControl.SelectedItem = null;
			subjectTreeControl.Subject = null;
			faceView.Face = null;
			icaoWarningView.Face = null;
			if (_newSubject.Status == NBiometricStatus.Ok)
			{
				var faces = _newSubject.Faces.ToArray();
				_newSubject.Clear();
				foreach (var face in faces)
				{
					_subject.Faces.Add(face);
				}
			}
			_newSubject = null;
			_subject = null;
			_biometricClient = null;

			SettingsManager.FacesMirrorHorizontally = chbMirrorHorizontally.Checked;

			base.OnNavigatingFrom();
		}

		#endregion

		#region Private methods

		private void SetIsBusy(bool value)
		{
			if (value)
				_isIdle.Reset();
			else
				_isIdle.Set();
		}

		private bool IsBusy()
		{
			return !_isIdle.WaitOne(0);
		}

		private void CancelAndWait()
		{
			if (IsBusy())
			{
				_biometricClient.Cancel();
				_isIdle.WaitOne();
			}
		}

		private void SetStatusText(Color backColor, string format, params object[] args)
		{
			lblStatus.Text = string.Format(format, args);
			lblStatus.BackColor = backColor;
			lblStatus.Visible = true;
		}

		private void EnableControls()
		{
			bool canCancel = rbFromCamera.Checked || rbFromVideo.Checked;
			bool isManual = chbManual.Checked;
			bool isStream = chbStream.Checked;
			bool isIdle = !IsBusy();
			bool isLocalCreate = (_biometricClient.LocalOperations & NBiometricOperations.CreateTemplate) == NBiometricOperations.CreateTemplate;
			bool checkIcao = chbCheckIcaoCompliance.Checked;
			chbStream.Enabled = !rbFromFile.Checked && isLocalCreate;
			chbStream.Checked = chbStream.Checked && isLocalCreate;
			chbManual.Enabled = !rbFromFile.Checked;
			chbCheckIcaoCompliance.Enabled = isIdle;
			rbFromCamera.Enabled = _biometricClient.FaceCaptureDevice != null;
			gbCaptureOptions.Enabled = isIdle;
			btnCancel.Enabled = !isIdle && canCancel;
			btnForceStart.Enabled = !isIdle && ((isManual && !_isExtractStarted) || checkIcao);
			btnForceEnd.Enabled = false;
			btnCancel.Visible = btnForceEnd.Visible = btnForceStart.Visible = canCancel;
			btnRepeat.Enabled = false;
			btnRepeat.Visible = chbWithGeneralization.Checked && btnCancel.Visible;

			bool boldStart = btnForceStart.Enabled && isManual;
			bool boldFinish = isIdle && _newSubject.Status == NBiometricStatus.Ok;
			btnForceStart.Font = new Font(btnForceStart.Font, boldStart ? FontStyle.Bold : FontStyle.Regular);
			btnForceStart.Text = checkIcao && !isManual ? "Force" : "Start";
			btnFinish.Font = new Font(btnFinish.Font, boldFinish ? FontStyle.Bold : FontStyle.Regular);

			busyIndicator.Visible = !isIdle;
		}

		private void ToggleRadioButton()
		{
			EnableControls();
		}

		private void OnFacePropertyChanged(object sender, PropertyChangedEventArgs e)
		{
			if (e.PropertyName == "Status")
			{
				BeginInvoke(new Action<NBiometricStatus>(OnFaceStatusChanged), ((NBiometric)sender).Status);
			}
		}

		private void OnFaceStatusChanged(NBiometricStatus status)
		{
			string format = _isExtractStarted ? "Extracion status: {0}" : "Detection status: {0}";
			Color backColor = status == NBiometricStatus.Ok || status == NBiometricStatus.None ? Color.Green : Color.Red;
			SetStatusText(backColor, _titlePrefix + format, status);
		}

		private void OnFaceCaptureDeviceChanged()
		{
			var device = _biometricClient.FaceCaptureDevice;
			if (device == null || !device.IsAvailable)
			{
				if (rbFromCamera.Checked) rbFromFile.Checked = true;
				rbFromCamera.Text = "From camera (Not connected)";
			}
			else
			{
				rbFromCamera.Text = string.Format("From camera ({0})", device.DisplayName);
			}
			EnableControls();
		}

		private void OnBiometricClientPropertyChanged(object sender, PropertyChangedEventArgs e)
		{
			if (e.PropertyName == "FaceCaptureDevice")
			{
				BeginInvoke(new Action(() =>
				{
					if (IsPageShown) OnFaceCaptureDeviceChanged();
				}));
			}
			else if (e.PropertyName == "CurrentBiometric")
			{
				BeginInvoke(new Action<NBiometric>(x =>
					{
						if (IsPageShown)
						{
							NFace face = (NFace)x;
							if (face != null) faceView.Face = face;
							if (face != null && chbCheckIcaoCompliance.Checked) icaoWarningView.Face = face;
							if (chbWithGeneralization.Checked && face != null)
							{
								int index = _newSubject.Faces.IndexOf(face);
								generalizationView.Selected = x;
								_titlePrefix = string.Format("Capturing face {0} of {1}. ", index + 1, SettingsManager.FacesGeneralizationRecordCount);
							}
							if (rbFromCamera.Checked || rbFromVideo.Checked)
							{
								_isExtractStarted = !chbManual.Checked;
								if (_currentBiometric != null) _currentBiometric.PropertyChanged -= OnFacePropertyChanged;
								_currentBiometric = face;
								if (_currentBiometric != null) _currentBiometric.PropertyChanged += OnFacePropertyChanged;
							}

							EnableControls();
						}
					}), ((NBiometricClient)sender).CurrentBiometric);
			}
		}

		private void UpdateWithTaskResult(NBiometricStatus status)
		{
			if (IsPageShown)
			{
				PrepareViews(false, chbCheckIcaoCompliance.Checked, status == NBiometricStatus.Ok);

				bool withGeneralization = chbWithGeneralization.Checked;
				Color backColor = status == NBiometricStatus.Ok ? Color.Green : Color.Red;
				SetStatusText(backColor, "Extraction status: {0}", status == NBiometricStatus.Timeout ? "Liveness check failed" : status.ToString());
				if (withGeneralization && status == NBiometricStatus.Ok)
				{
					NFace generalized = _newSubject.Faces.Last();
					generalizationView.Generalized = new[] { generalized };
					generalizationView.Selected = generalized;
				}
				generalizationView.EnableMouseSelection = true;
				EnableControls();
			}
		}

		private void OnCurrentBiometricCompleted(object sender, EventArgs e)
		{
			NFace face = (NFace)_biometricClient.CurrentBiometric;
			NBiometricStatus status = face.Status;
			if (status == NBiometricStatus.Ok)
			{
				var attributes = face.Objects.FirstOrDefault();
				var child = attributes != null ? attributes.Child : null;
				if (child != null && child.Status != NBiometricStatus.Ok)
					status = child.Status;
			}

			BeginInvoke(new Action<NBiometricStatus>(OnCurrentBiometricCompletedInternal), status);
		}

		private void OnCurrentBiometricCompletedInternal(NBiometricStatus status)
		{
			bool allowRepeat = btnRepeat.Visible && status != NBiometricStatus.Ok;
			if (!allowRepeat)
				_biometricClient.Force();
			else
			{
				SetStatusText(Color.Red, _titlePrefix + "Extracion status: {0}", status);
			}
			btnRepeat.Enabled = allowRepeat;
		}

		private void ChbCheckIcaoComplianceCheckedChanged(object sender, EventArgs e)
		{
			if (chbCheckIcaoCompliance.Checked)
			{
				chbStream.Checked = chbStream.Enabled;
				chbManual.Checked = false;
			}
		}

		#endregion

		#region Private form events

		private void BtnForceStartClick(object sender, EventArgs e)
		{
			if (!chbCheckIcaoCompliance.Checked)
			{
				_isExtractStarted = true;
				btnForceEnd.Enabled = chbStream.Checked;
				btnForceStart.Enabled = false;
			}
			else if (chbManual.Checked)
			{
				btnForceStart.Text = "Force";
				btnForceStart.Font = new Font(btnFinish.Font, FontStyle.Regular);
			}
			_biometricClient.Force();
		}

		private void BtnForceEndClick(object sender, EventArgs e)
		{
			btnForceEnd.Enabled = false;
			_isExtractStarted = false;
			_biometricClient.Force();
		}

		private void BtnCancelClick(object sender, EventArgs e)
		{
			_biometricClient.Cancel();
		}

		private void RadioButtonCheckedChanged(object sender, EventArgs e)
		{
			if (((RadioButton)sender).Checked)
				ToggleRadioButton();
		}

		private void BtnFinishClick(object sender, EventArgs e)
		{
			PageController.NavigateToStartPage();
		}

		private void PrepareViews(bool isCapturing, bool checkIcao)
		{
			PrepareViews(isCapturing, checkIcao, true);
		}

		private void PrepareViews(bool isCapturing, bool checkIcao, bool isOk)
		{
			icaoWarningView.Visible = checkIcao;
			if (isCapturing)
			{
				faceView.ShowAge = !checkIcao;
				faceView.ShowEmotions = !checkIcao;
				faceView.ShowExpression = !checkIcao;
				faceView.ShowGender = !checkIcao;
				faceView.ShowProperties = !checkIcao;
				faceView.ShowIcaoArrows = true;
				subjectTreeControl.Visible = false;
			}
			else
			{
				faceView.ShowAge = true;
				faceView.ShowEmotions = true;
				faceView.ShowExpression = true;
				faceView.ShowGender = true;
				faceView.ShowProperties = true;
				faceView.ShowIcaoArrows = false;
				subjectTreeControl.Visible = checkIcao && isOk;
				if (checkIcao)
				{
					subjectTreeControl.UpdateTree();
					var node = subjectTreeControl.Nodes.First();
					subjectTreeControl.SelectedItem = node.GetChildren().FirstOrDefault() ?? node;
				}
			}
			faceView.Invalidate();
		}

		private void BtnCaptureClick(object sender, EventArgs e)
		{
			bool generalize = chbWithGeneralization.Checked;
			bool fromFile = rbFromFile.Checked;
			bool fromCamera = rbFromCamera.Checked;
			bool checkIcao = chbCheckIcaoCompliance.Checked;
			int count = generalize ? SettingsManager.FacesGeneralizationRecordCount : 1;
			NBiometricCaptureOptions options = NBiometricCaptureOptions.None;
			if (chbManual.Checked) options |= NBiometricCaptureOptions.Manual;
			if (chbStream.Checked) options |= NBiometricCaptureOptions.Stream;

			lblStatus.Visible = false;
			_titlePrefix = string.Empty;
			_newSubject.Clear();
			faceView.Face = null;
			generalizationView.Clear();
			generalizationView.EnableMouseSelection = false;
			generalizationView.Visible = generalize;

			List<string> selectedFiles = new List<string>();
			string title = fromFile ? "Select image" : "Select video file";
			string titleFormat = fromFile ? "Select face image ({0} out of {1})" : "Select video file ({0} out of {1})";
			string fileFilter = fromFile ? NImages.GetOpenFileFilterString(true, true) : null;
			if (rbFromFile.Checked || rbFromVideo.Checked)
			{
				openFileDialog.FileName = null;
				openFileDialog.Filter = fileFilter;
				openFileDialog.Title = title;
				while (selectedFiles.Count < count)
				{
					if (generalize) openFileDialog.Title = string.Format(titleFormat, selectedFiles.Count + 1, count);
					if (openFileDialog.ShowDialog() != DialogResult.OK) return;
					selectedFiles.Add(openFileDialog.FileName);
				}
			}

			int id = generalize ? _sessionId : -1;
			for (int i = 0; i < count; i++)
			{
				NFace face = new NFace
				{
					SessionId = id,
					FileName = !fromCamera ? selectedFiles[i] : null,
					CaptureOptions = options
				};
				_newSubject.Faces.Add(face);
			}
			faceView.Face = _newSubject.Faces.First();

			if (generalize)
			{
				generalizationView.Biometrics = _newSubject.Faces.ToArray();
				generalizationView.Selected = _newSubject.Faces.First();
			}

			icaoWarningView.Face = _newSubject.Faces.First();

			_biometricClient.FacesCheckIcaoCompliance = checkIcao;
			NBiometricOperations operations = fromFile ? NBiometricOperations.CreateTemplate : NBiometricOperations.Capture | NBiometricOperations.CreateTemplate;
			if (checkIcao) operations |= NBiometricOperations.Segment;
			NBiometricTask biometricTask = _biometricClient.CreateTask(operations, _newSubject);
			SetStatusText(Color.Orange, fromFile ? "Extracting template ..." : "Starting capturing ...");
			SetIsBusy(true);
			_biometricClient.BeginPerformTask(biometricTask, OnCreateTemplateCompleted, null);
			EnableControls();
			PrepareViews(true, checkIcao);
		}

		private void OnCreateTemplateCompleted(IAsyncResult r)
		{
			NBiometricStatus status = NBiometricStatus.InternalError;
			try
			{
				var biometricTask = _biometricClient.EndPerformTask(r);
				status = biometricTask.Status;
				if (biometricTask.Error != null) Utilities.ShowError(biometricTask.Error);
			}
			catch (Exception ex)
			{
				Utilities.ShowError(ex);
			}
			finally
			{
				SetIsBusy(false);
				BeginInvoke(new Action<NBiometricStatus>(UpdateWithTaskResult), status);
			}
		}

		private void BtnRepeatClick(object sender, EventArgs e)
		{
			_biometricClient.Repeat();
		}

		private void CaptureFacePageLoad(object sender, EventArgs e)
		{
			chbStream.Checked = (_biometricClient.LocalOperations & NBiometricOperations.CreateTemplate) != 0;
		}

		private void ChbMirrorHorizontallyCheckedChanged(object sender, EventArgs e)
		{
			faceView.MirrorHorizontally = chbMirrorHorizontally.Checked;
			faceView.Invalidate();
		}

		private void OnSubjectTreeControlPropertyChanged(object sender, PropertyChangedEventArgs e)
		{
			if (subjectTreeControl.Visible && e.PropertyName == "SelectedItem")
			{
				var selected = subjectTreeControl.SelectedItem;
				if (selected != null)
				{
					if (generalizationView.Visible)
					{
						generalizationView.Biometrics = selected.Items;
						generalizationView.Generalized = selected.GetAllGeneralized();
						generalizationView.Selected = selected.AllItems.First();
					}
					else
					{
						faceView.Face = selected.Items.First() as NFace;
						icaoWarningView.Face = faceView.Face;
					}
				}
			}
		}

		#endregion
	}
}

File: java/Samples/Biometrics/VB/AbisSampleVB/SubjectEditor/CaptureFacePage.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Drawing
Imports System.Linq
Imports System.Threading
Imports System.Windows.Forms
Imports Neurotec.Biometrics
Imports Neurotec.Biometrics.Client
Imports Neurotec.Images

Partial Public Class CaptureFacePage
	Inherits PageBase
#Region "Public constructor"

	Public Sub New()
		InitializeComponent()
		AddHandler rbFromVideo.CheckedChanged, AddressOf RadioButtonCheckedChanged
		AddHandler rbFromFile.CheckedChanged, AddressOf RadioButtonCheckedChanged
		AddHandler rbFromCamera.CheckedChanged, AddressOf RadioButtonCheckedChanged
	End Sub

#End Region

#Region "Private fields"

	Private _subject As NSubject
	Private _biometricClient As NBiometricClient
	Private _isIdle As New ManualResetEvent(True)
	Private _currentBiometric As NFace = Nothing
	Private _newSubject As NSubject
	Private _isExtractStarted As Boolean = False
	Private _sessionId As Integer = -1
	Private _titlePrefix As String = String.Empty

#End Region

#Region "Public overridden methods"

	Public Overrides Sub OnNavigatedTo(ByVal ParamArray args() As Object)
		If args Is Nothing OrElse args.Length <> 2 Then
			Throw New ArgumentException("args")
		End If

		_subject = CType(args(0), NSubject)
		_biometricClient = CType(args(1), NBiometricClient)
		If _subject Is Nothing OrElse _biometricClient Is Nothing Then
			Throw New ArgumentException("args")
		End If

		_sessionId = If(_subject.Faces.Count > 0, _subject.Faces.Max(Function(x) x.SessionId) + 1, 0)
		_newSubject = New NSubject()
		AddHandler _biometricClient.PropertyChanged, AddressOf OnBiometricClientPropertyChanged
		_biometricClient.CurrentBiometricCompletedTimeout = 5000
		AddHandler _biometricClient.CurrentBiometricCompleted, AddressOf OnCurrentBiometricCompleted

		subjectTreeControl.Visible = False
		icaoWarningView.Visible = False
		OnFaceCaptureDeviceChanged()
		ToggleRadioButton()
		lblStatus.Visible = False
		generalizationView.Visible = False

		Dim mirrorHorizontally As Boolean = SettingsManager.FacesMirrorHorizontally
		chbMirrorHorizontally.Checked = mirrorHorizontally
		faceView.MirrorHorizontally = mirrorHorizontally

		subjectTreeControl.Subject = _newSubject

		MyBase.OnNavigatedTo(args)
	End Sub

	Public Overrides Sub OnNavigatingFrom()
		If IsBusy() Then
			LongActionDialog.ShowDialog(Me, "Finish current action ...", AddressOf CancelAndWait)
		End If
		RemoveHandler _biometricClient.PropertyChanged, AddressOf OnBiometricClientPropertyChanged
		RemoveHandler _biometricClient.CurrentBiometricCompleted, AddressOf OnCurrentBiometricCompleted
		_biometricClient.CurrentBiometricCompletedTimeout = 0
		subjectTreeControl.SelectedItem = Nothing
		subjectTreeControl.Subject = Nothing
		faceView.Face = Nothing
		icaoWarningView.Face = Nothing
		If _newSubject.Status = NBiometricStatus.Ok Then
			Dim faces = _newSubject.Faces.ToArray()
			_newSubject.Clear()
			For Each face In faces
				_subject.Faces.Add(face)
			Next face
		End If
		_newSubject = Nothing
		_subject = Nothing
		_biometricClient = Nothing

		SettingsManager.FacesMirrorHorizontally = chbMirrorHorizontally.Checked

		MyBase.OnNavigatingFrom()
	End Sub

#End Region

#Region "Private methods"

	Private Sub SetIsBusy(ByVal value As Boolean)
		If value Then
			_isIdle.Reset()
		Else
			_isIdle.Set()
		End If
	End Sub

	Private Function IsBusy() As Boolean
		Return Not _isIdle.WaitOne(0)
	End Function

	Private Sub CancelAndWait()
		If IsBusy() Then
			_biometricClient.Cancel()
			_isIdle.WaitOne()
		End If
	End Sub

	Private Sub SetStatusText(ByVal backColor As Color, ByVal format As String, ByVal ParamArray args() As Object)
		lblStatus.Text = String.Format(format, args)
		lblStatus.BackColor = backColor
		lblStatus.Visible = True
	End Sub

	Private Sub EnableControls()
		Dim canCancel As Boolean = rbFromCamera.Checked OrElse rbFromVideo.Checked
		Dim isManual As Boolean = chbManual.Checked
		Dim isStream As Boolean = chbStream.Checked
		Dim isIdle As Boolean = Not IsBusy()
		Dim isLocalCreate = (_biometricClient.LocalOperations And NBiometricOperations.CreateTemplate) = NBiometricOperations.CreateTemplate
		Dim checkIcao As Boolean = chbCheckIcaoCompliance.Checked
		chbStream.Enabled = (Not rbFromFile.Checked) AndAlso isLocalCreate
		chbStream.Checked = chbStream.Checked AndAlso isLocalCreate
		chbManual.Enabled = Not rbFromFile.Checked
		chbCheckIcaoCompliance.Enabled = isIdle
		rbFromCamera.Enabled = _biometricClient.FaceCaptureDevice IsNot Nothing
		gbCaptureOptions.Enabled = isIdle
		btnCancel.Enabled = (Not isIdle) AndAlso canCancel
		btnForceStart.Enabled = (Not isIdle) AndAlso ((isManual AndAlso (Not _isExtractStarted)) OrElse checkIcao)
		btnForceEnd.Enabled = False
		btnForceStart.Visible = canCancel
		btnForceEnd.Visible = btnForceStart.Visible
		btnCancel.Visible = btnForceEnd.Visible
		btnRepeat.Enabled = False
		btnRepeat.Visible = chbWithGeneralization.Checked AndAlso btnCancel.Visible

		Dim boldStart As Boolean = btnForceStart.Enabled AndAlso isManual
		Dim boldFinish As Boolean = isIdle AndAlso _newSubject.Status = NBiometricStatus.Ok
		btnForceStart.Font = New Font(btnForceStart.Font, If(boldStart, FontStyle.Bold, FontStyle.Regular))
		btnForceStart.Text = If(checkIcao AndAlso (Not isManual), "Force", "Start")
		btnFinish.Font = New Font(btnFinish.Font, If(boldFinish, FontStyle.Bold, FontStyle.Regular))

		busyIndicator.Visible = Not isIdle
	End Sub

	Private Sub ToggleRadioButton()
		EnableControls()
	End Sub

	Private Sub OnFacePropertyChanged(ByVal sender As Object, ByVal e As PropertyChangedEventArgs)
		If e.PropertyName = "Status" Then
			BeginInvoke(New Action(Of NBiometricStatus)(AddressOf OnFaceStatusChanged), (CType(sender, NBiometric)).Status)
		End If
	End Sub

	Private Sub OnFaceStatusChanged(ByVal status As NBiometricStatus)
		Dim format As String = If(_isExtractStarted, "Extracion status: {0}", "Detection status: {0}")
		Dim backColor As Color = If(status = NBiometricStatus.Ok OrElse status = NBiometricStatus.None, Color.Green, Color.Red)
		SetStatusText(backColor, _titlePrefix & format, status)
	End Sub

	Private Sub OnFaceCaptureDeviceChanged()
		Dim device = _biometricClient.FaceCaptureDevice
		If device Is Nothing OrElse (Not device.IsAvailable) Then
			If rbFromCamera.Checked Then
				rbFromFile.Checked = True
			End If
			rbFromCamera.Text = "From camera (Not connected)"
		Else
			rbFromCamera.Text = String.Format("From camera ({0})", device.DisplayName)
		End If
		EnableControls()
	End Sub

	Private Sub OnBiometricClientPropertyChanged(ByVal sender As Object, ByVal e As PropertyChangedEventArgs)
		If e.PropertyName = "FaceCaptureDevice" Then
			BeginInvoke(New Action(AddressOf InvokeOnFaceCaptureDeviceChanged))
		ElseIf e.PropertyName = "CurrentBiometric" Then
			BeginInvoke(New Action(Of NBiometric)(AddressOf OnCurrentBiometricChanged), (CType(sender, NBiometricClient)).CurrentBiometric)
		End If
	End Sub
	Private Sub InvokeOnFaceCaptureDeviceChanged()
		If IsPageShown Then
			OnFaceCaptureDeviceChanged()
		End If
	End Sub
	Private Sub OnCurrentBiometricChanged(ByVal x As Object)
		If IsPageShown Then
			Dim face As NFace = CType(x, NFace)
			If face IsNot Nothing Then
				faceView.Face = face
			End If
			If face IsNot Nothing AndAlso chbCheckIcaoCompliance.Checked Then
				icaoWarningView.Face = face
			End If
			If chbWithGeneralization.Checked AndAlso face IsNot Nothing Then
				Dim index As Integer = _newSubject.Faces.IndexOf(face)
				generalizationView.Selected = x
				_titlePrefix = String.Format("Capturing face {0} of {1}. ", index + 1, SettingsManager.FacesGeneralizationRecordCount)
			End If
			If rbFromCamera.Checked OrElse rbFromVideo.Checked Then
				_isExtractStarted = Not chbManual.Checked
				If _currentBiometric IsNot Nothing Then
					RemoveHandler _currentBiometric.PropertyChanged, AddressOf OnFacePropertyChanged
				End If
				_currentBiometric = face
				If _currentBiometric IsNot Nothing Then
					AddHandler _currentBiometric.PropertyChanged, AddressOf OnFacePropertyChanged
				End If
			End If
			EnableControls()
		End If
	End Sub

	Private Sub UpdateWithTaskResult(ByVal status As NBiometricStatus)
		If IsPageShown Then
			PrepareViews(False, chbCheckIcaoCompliance.Checked, status = NBiometricStatus.Ok)

			Dim withGeneralization As Boolean = chbWithGeneralization.Checked
			Dim backColor As Color = If(status = NBiometricStatus.Ok, Color.Green, Color.Red)
			SetStatusText(backColor, "Extraction status: {0}", If(status = NBiometricStatus.Timeout, "Liveness check failed", status.ToString()))
			If withGeneralization AndAlso status = NBiometricStatus.Ok Then
				Dim generalized As NFace = _newSubject.Faces.Last()
				generalizationView.Generalized = New NFace() {generalized}
				generalizationView.Selected = generalized
			End If
			generalizationView.EnableMouseSelection = True
			EnableControls()
		End If
	End Sub

	Private Sub OnCurrentBiometricCompleted(ByVal sender As Object, ByVal e As EventArgs)
		Dim face As NFace = CType(_biometricClient.CurrentBiometric, NFace)
		Dim status As NBiometricStatus = face.Status
		If status = NBiometricStatus.Ok Then
			Dim attributes = face.Objects.FirstOrDefault()
			Dim child = If(attributes IsNot Nothing, attributes.Child, Nothing)
			If child IsNot Nothing AndAlso child.Status <> NBiometricStatus.Ok Then
				status = child.Status
			End If
		End If

		BeginInvoke(New Action(Of NBiometricStatus)(AddressOf OnCurrentBiometricCompletedInternal), status)
	End Sub

	Private Sub OnCurrentBiometricCompletedInternal(ByVal status As Object)
		Dim allowRepeat As Boolean = btnRepeat.Visible AndAlso status <> NBiometricStatus.Ok
		If (Not allowRepeat) Then
			_biometricClient.Force()
		Else
			SetStatusText(Color.Red, _titlePrefix & "Extracion status: {0}", status)
		End If
		btnRepeat.Enabled = allowRepeat
	End Sub

	Private Sub PrepareViews(ByVal isCapturing As Boolean, ByVal checkIcao As Boolean)
		PrepareViews(isCapturing, checkIcao, True)
	End Sub

	Private Sub PrepareViews(ByVal isCapturing As Boolean, ByVal checkIcao As Boolean, ByVal isOk As Boolean)
		icaoWarningView.Visible = checkIcao
		If isCapturing Then
			faceView.ShowAge = Not checkIcao
			faceView.ShowEmotions = Not checkIcao
			faceView.ShowExpression = Not checkIcao
			faceView.ShowGender = Not checkIcao
			faceView.ShowProperties = Not checkIcao
			faceView.ShowIcaoArrows = True
			subjectTreeControl.Visible = False
		Else
			faceView.ShowAge = True
			faceView.ShowEmotions = True
			faceView.ShowExpression = True
			faceView.ShowGender = True
			faceView.ShowProperties = True
			faceView.ShowIcaoArrows = False
			subjectTreeControl.Visible = checkIcao AndAlso isOk
			If checkIcao Then
				subjectTreeControl.UpdateTree()
				Dim node = subjectTreeControl.Nodes.First()
				subjectTreeControl.SelectedItem = If(node.GetChildren().FirstOrDefault(), node)
			End If
		End If
		faceView.Invalidate()
	End Sub

#End Region

#Region "Private form events"

	Private Sub BtnForceStartClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnForceStart.Click
		If (Not chbCheckIcaoCompliance.Checked) Then
			_isExtractStarted = True
			btnForceEnd.Enabled = chbStream.Checked
			btnForceStart.Enabled = False
		ElseIf chbManual.Checked Then
			btnForceStart.Text = "Force"
			btnForceStart.Font = New Font(btnFinish.Font, FontStyle.Regular)
		End If
		_biometricClient.Force()
	End Sub

	Private Sub BtnForceEndClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnForceEnd.Click
		btnForceEnd.Enabled = False
		_isExtractStarted = False
		_biometricClient.Force()
	End Sub

	Private Sub BtnCancelClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnCancel.Click
		_biometricClient.Cancel()
	End Sub

	Private Sub RadioButtonCheckedChanged(ByVal sender As Object, ByVal e As EventArgs)
		If (CType(sender, RadioButton)).Checked Then
			ToggleRadioButton()
		End If
	End Sub

	Private Sub BtnFinishClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnFinish.Click
		PageController.NavigateToStartPage()
	End Sub

	Private Sub BtnCaptureClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnCapture.Click
		Dim generalize As Boolean = chbWithGeneralization.Checked
		Dim fromFile As Boolean = rbFromFile.Checked
		Dim fromCamera As Boolean = rbFromCamera.Checked
		Dim checkIcao As Boolean = chbCheckIcaoCompliance.Checked
		Dim count As Integer = If(generalize, SettingsManager.FacesGeneralizationRecordCount, 1)
		Dim options As NBiometricCaptureOptions = NBiometricCaptureOptions.None
		If chbManual.Checked Then
			options = options Or NBiometricCaptureOptions.Manual
		End If
		If chbStream.Checked Then
			options = options Or NBiometricCaptureOptions.Stream
		End If

		lblStatus.Visible = False
		_titlePrefix = String.Empty
		_newSubject.Clear()
		faceView.Face = Nothing
		generalizationView.Clear()
		generalizationView.EnableMouseSelection = False
		generalizationView.Visible = generalize

		Dim selectedFiles As New List(Of String)()
		Dim title As String = If(fromFile, "Select image", "Select video file")
		Dim titleFormat As String = If(fromFile, "Select face image ({0} out of {1})", "Select video file ({0} out of {1})")
		Dim fileFilter As String = If(fromFile, NImages.GetOpenFileFilterString(True, True), Nothing)
		If rbFromFile.Checked OrElse rbFromVideo.Checked Then
			openFileDialog.FileName = Nothing
			openFileDialog.Filter = fileFilter
			openFileDialog.Title = title
			Do While selectedFiles.Count < count
				If generalize Then
					openFileDialog.Title = String.Format(titleFormat, selectedFiles.Count + 1, count)
				End If
				If openFileDialog.ShowDialog() <> DialogResult.OK Then
					Return
				End If
				selectedFiles.Add(openFileDialog.FileName)
			Loop
		End If

		Dim id As Integer = If(generalize, _sessionId, -1)
		For i As Integer = 0 To count - 1
			Dim face As NFace = New NFace With {.SessionId = id, .FileName = If((Not fromCamera), selectedFiles(i), Nothing), .CaptureOptions = options}
			_newSubject.Faces.Add(face)
		Next i
		faceView.Face = _newSubject.Faces.First()

		If generalize Then
			generalizationView.Biometrics = _newSubject.Faces.ToArray()
			generalizationView.Selected = _newSubject.Faces.First()
		End If

		icaoWarningView.Face = _newSubject.Faces.First()

		_biometricClient.FacesCheckIcaoCompliance = checkIcao
		Dim operations As NBiometricOperations = If(fromFile, NBiometricOperations.CreateTemplate, NBiometricOperations.Capture Or NBiometricOperations.CreateTemplate)
		If checkIcao Then
			operations = operations Or NBiometricOperations.Segment
		End If

		Dim biometricTask As NBiometricTask = _biometricClient.CreateTask(operations, _newSubject)
		SetStatusText(Color.Orange, If(fromFile, "Extracting template ...", "Starting capturing ..."))
		SetIsBusy(True)
		_biometricClient.BeginPerformTask(biometricTask, AddressOf OnCreateTemplateCompleted, Nothing)

		EnableControls()
		PrepareViews(True, checkIcao)
	End Sub

	Private Sub OnCreateTemplateCompleted(ByVal r As IAsyncResult)
		Dim status As NBiometricStatus = NBiometricStatus.InternalError
		Try
			Dim biometricTask = _biometricClient.EndPerformTask(r)
			status = biometricTask.Status
			If biometricTask.Error IsNot Nothing Then
				Utilities.ShowError(biometricTask.Error)
			End If
		Catch ex As Exception
			Utilities.ShowError(ex)
		Finally
			SetIsBusy(False)
			BeginInvoke(New Action(Of NBiometricStatus)(AddressOf UpdateWithTaskResult), status)
		End Try
	End Sub

	Private Sub BtnRepeatClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnRepeat.Click
		_biometricClient.Repeat()
	End Sub

	Private Sub CaptureFacePageLoad(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
		chbStream.Checked = (_biometricClient.LocalOperations And NBiometricOperations.CreateTemplate) <> 0
	End Sub

	Private Sub ChbMirrorHorizontallyCheckedChanged(ByVal sender As Object, ByVal e As EventArgs) Handles chbMirrorHorizontally.CheckedChanged
		faceView.MirrorHorizontally = chbMirrorHorizontally.Checked
		faceView.Invalidate()
	End Sub

	Private Sub ChbCheckIcaoComplianceCheckedChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles chbCheckIcaoCompliance.CheckedChanged
		If chbCheckIcaoCompliance.Checked Then
			chbStream.Checked = chbStream.Enabled
			chbManual.Checked = False
		End If
	End Sub

	Private Sub SubjectTreeControlPropertyChanged(ByVal sender As System.Object, ByVal e As System.ComponentModel.PropertyChangedEventArgs) Handles subjectTreeControl.PropertyChanged
		If subjectTreeControl.Visible AndAlso e.PropertyName = "SelectedItem" Then
			Dim selected = subjectTreeControl.SelectedItem
			If selected IsNot Nothing Then
				If generalizationView.Visible Then
					generalizationView.Biometrics = selected.Items
					generalizationView.Generalized = selected.GetAllGeneralized()
					generalizationView.Selected = selected.AllItems.First()
				Else
					faceView.Face = TryCast(selected.Items.First(), NFace)
					icaoWarningView.Face = faceView.Face
				End If
			End If
		End If
	End Sub

#End Region
End Class
